### ISSUE TITLE
Question: Very low throughput (250 FPS/sec)

### ISSUE THREAD

**Issue by steffenvan** [jaimeyzzz/impala_horovod_gym#4]
Hi, 

I'm wondering how you managed to replicate the throughput in the original IMPALA? Currently I'm trying to replicate it myself in a multi-task learning setup (using atari), but I only see a throughput on roughly ~250 frames/sec (1k/frames if the action repeats=4). 

I'm using a
- Ubuntu 18.04.2
- GeForce GTX 1080 
- 28 CPU Intel Xeon E5-2620 v4 (8 cores each) 
I'm currently running with 16 actors and a batch size on 8 (as this empirically gave the highest throughput). 

Do you have a suggestion on what might have gone wrong, or if you at some point faced similar issues? 

If you need more information, please let me know!

**Comment by pengsun** [comment]
Hi, the throughput we replicate with this repo is described here
https://github.com/loyave/impala_horovod_gym#case-studies
and here
https://github.com/loyave/impala_horovod_gym/blob/dev-horovod/sandbox/speed.md

We use standalone CPU machines (remotes across the network) as dedicated actors, and the learners are M40 GPU machines.

Actor is producer while learner is consumer, our observation is that the data producing speed can be the bottleneck. Use more actors (hundreds to thousands) if possible.

**Comment by steffenvan** [comment]
Thanks for the quick reply! 

So 2 quick questions just to clarify:
 
1. Every actor has access to 1 CPU, if I understand correctly? If so, have you tried to allocate multiple actors on one CPU, and whether it lead to better/worse results? 

2. The speed that you report in [speed.md]( https://github.com/loyave/impala_horovod_gym/blob/dev-horovod/sandbox/speed.md), is that after or before you have multiplied the amount of frames/sec with 4?

**Comment by pengsun** [comment]
Hi,
1. I checked the log and found that we used 4 CPU cores per actor. Surely, too few CPU cores per actor would decrease the speed (as an actor must do a couple of things: manage the env interaction, do the NN inference, etc.), and too many is wasteful. Four is just an "arbitrary number", we didn't do a strict test to decide an optimal no. CPU cores. 
2. It's BEFORE we multiply 4. E.g.,  "1 learner 15 actors", 2.3K fps. Then the environment frame FPS is 2.3k x 4 = 7.2k
See our detailed description in https://github.com/loyave/impala_horovod_gym/blob/dev-horovod/sandbox/speed.md

**Comment by steffenvan** [comment]
Hi again!

Apologize the belated response, I had to configure my setup as I just got access to more CPU's (64) and better GPU (Tesla V100) with which I just run a vanilla IMPALA agent in DMLab30 and received similar results as reported in the original paper.

But doing the same with the Atari environment still yields a very low throughput (or at least reported in the tensorflow logging). I suspect it has to do with the environment setup. I have tried to match it to the way you're doing it - but am still reported a low throughput and an agent who basically doesn't learn much. I suspect it might have to with two things I currently haven't included from your setup, to which I have two questions (perhaps this should be opened in another issue, let me know if that's the case):  

1. What is the purpose of the additional `sum_raw_reward` and `sum_raw_step` in
[https://github.com/loyave/impala_horovod_gym/blob/dev-horovod/environments.py#L68](https://github.com/loyave/impala_horovod_gym/blob/dev-horovod/environments.py#L68) is?
 
2. And why did you add an additional `NoRwdWrapper`. Was it just to make the agent converge faster? I might be wrong, but I don't believe any papers have reported doing a similar thing. 

Thank you very much!

**Comment by pengsun** [comment]
1. it's for logging purpose. The `info` field is usually used for any interested variable that is not directly related to the training logics.
2. don't worry, it's just our internal random trial. The purpose is similar to `NoopResetEnv`. It's fine to use either one, and no significant difference is observed in my impression...

**Comment by steffenvan** [comment]
Okay thanks for the reply! 

Sorry that I keep extending this, but I have now tried to run (a single machine setup) that sets up the environment as you do with the Atari environment.  And comparomg the throughput with the vanilla impala implementation with DMlab30, it appears that the vanilla impala implementation has a much higher throughput (~14k-16k frames/sec) than the Atari environment (only ~800 frames/sec). Is it something that you encountered as well?

**Comment by pengsun** [comment]
it's fine:) 

We didn't try the single machine setup, but for the distributed remote machine setup the throughput can be approximately reproduced with Atari env after carefully noticing a couple of technical details, including the environment as TF node trick(the PyGymProcess class in this repo), the NN structure, the GPU type (vanilla IMPALA used a V100 in my impression), whether to x4 for the frames/sec measure, etc.

**Comment by steffenvan** [comment]
I have created a similar class to your PyGymProcess, with the same neural network structure as the feed-forward network (that's also used in DQN and A3C) and I am using a V100 coupled with 64 CPU's (on a single machine though). Might there be another bottleneck I have overlooked?

**Comment by pengsun** [comment]
That seems a thorough reproduction... No idea why... Perhaps you could refer to the profiling tool as suggested by the authors of the original `scalable_agent:
https://github.com/deepmind/scalable_agent/issues/14#issuecomment-409207996

**Comment by steffenvan** [comment]
Just an update on the situation: after many hours of debugging I realised the low throughput was caused by setting the local device to the **GPU** instead of **CPU** (for unknown reasons). The specific part of the code would correspond to this line: https://github.com/loyave/impala_horovod_gym/blob/dev-horovod/experiment.py#L381 (where it is, as correctly written in the code, the CPU). 

I'm therefore closing this issue - thanks a lot for the help!
